baseplate/diagnostics/tracing.py

```diff
@@ -0,0 +1,277 @@
+        child_span.register(TraceSpanObserver(self.recorder))
+
+    def on_finish(self, exc_info):
+        self.end = current_epoch_microseconds()
+        self.elapsed = self.end - self.start
+        self.record()
+
+    def _serialize(self):
+        """Serialize span information into Zipkin-accepted format."""
+        annotations = []
+
+        annotations.append(
+            self._create_time_annotation(
+                ANNOTATIONS['SERVER_RECEIVE'],
+                self.start,
+            )
+        )
+        annotations.append(
+            self._create_time_annotation(
+                ANNOTATIONS['SERVER_SEND'],
+                self.end,
+            )
+        )
+
+        return self._to_span_obj(annotations, [])
+
+
+class NullRecorder(object):
+    def __init__(self):
+        self.logger = logging.getLogger(self.__class__.__name__)
+
+    def send(self, span):
+        """Write a set of spans to debug log."""
+        self.logger.debug("Span recording for trace_id=%s: %s",
+                          span['traceId'],
+                          span)
+
+
+class RemoteRecorder(object):
+    """Interface for recording spans to a remote collector.
+
+    The RemoteRecorder adds spans to an in-memory Queue for a background
+    thread worker to process. It currently does not shut down gracefully -
+    in the event of parent process exit, any remaining spans will be discarded.
+    """
+    def __init__(self, endpoint, max_conns, max_queue_size=10000):
+        self.logger = logging.getLogger(self.__class__.__name__)
+
+        adapter = requests.adapters.HTTPAdapter(pool_connections=max_conns,
+                                                pool_maxsize=max_conns)
+        self.session = requests.Session()
+        self.session.mount('http://', adapter)
+        self.endpoint = "http://%s/api/v1/spans" % endpoint
+
+        self.span_queue = queue.Queue(maxsize=max_queue_size)
+        self.flush_worker = None
+        self.flush_worker = threading.Thread(target=self._flush_spans)
+        self.flush_worker.name = "span remote recorder"
+        self.flush_worker.daemon = True
+        self.flush_worker.start()
+
+    def _flush_spans(self):
+        """Send a set of spans to remote collector.
+
+        This reads a batch of at most 10 spans off the recorder queue
+        and sends them to a remote recording endpoint. If the queue
+        empties while being processed before reaching 10 spans, we flush
+        immediately.
+        """
+        while True:
+            spans = []
+            try:
+                while len(spans) < 10:
+                    spans.append(self.span_queue.get(False, timeout=.1))
+            except queue.Empty:
+                pass
+            finally:
+                if spans:
+                    recording_req = requests.Request(
+                        'POST', self.endpoint,
+                        data=json.dumps(spans),
```

Comment:
:no_bell: there's also a `json` kwarg that does encoding for you. http://docs.python-requests.org/en/master/user/quickstart/#more-complicated-post-requests